Development Mode with multiple maven modules not working

**Describe the bug**
I'm not sure if this is a genuine issue or a known limitation.
I also checked in the doc but I couldn't find any reference to it.

I've tried to re-create our production project module organisation in a mini project on github:
https://github.com/uberto/h4kcluster-quarkus

Now we have different modules for domain, adapter, ui and the actual applications.

We don't have any source on the parent project.

Maven build is fine and when I start quarkus with dev mode specifying "launcher" (the project with the servlet as quarkus entry point) it starts correctly:
./mvnw -pl launcher quarkus:dev

It will also detect and hot-reload the changes I do to the launcher project, but it is not detecting if I change something else. I would need to stop, "mvn install" and restart to detect those.

Am I missing something?

Would it be better using Gradle? since it has direct project dependencies and doesn't require an install?


**Expected behavior**
Changing a submodule, for example "domain" should be reflected immediately when running quarkus in dev mode

**Actual behavior**
I need to stop dev mode and mvn install to see the change


**To Reproduce**
Steps to reproduce the behavior:
1. clone https://github.com/uberto/h4kcluster-quarkus
2. ./mvnw -pl launcher quarkus:dev
3. try to change something in domain package and see the it's reflected in the running app


**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Linux pop-os 5.4.0-7634-generic #38~1592497129~20.04~9a1ea2e-Ubuntu SMP Fri Jun 19 22:43:37 UTC  x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: 
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.20.0, JRE 14 Linux amd64-64-Bit Compressed References 20200416_44 (JIT enabled, AOT enabled)
OpenJ9   - 05fa2d361
OMR      - d4365f371
JCL      - 5757187cae based on jdk-14.0.1+7)

 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 
1.6.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Maven:  Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/ubertobarbini/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 14.0.1, vendor: AdoptOpenJDK, runtime: /home/ubertobarbini/.sdkman/candidates/java/14.0.1.j9-adpt
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.0-7634-generic", arch: "amd64", family: "unix"

I realized I was using J9 for these tests, going back to OpenJdk 14 gave me better timings.

the first hot-replace is still quite slow (1.5 s.) but next call is 0.5s and then progressively faster until it stabilizes on 320 ms.

I generated the svg file from the profile agent (btw nice improvement on old way to have flamegraphs).
Is there anything specific I should look at it?



Btw about the launching:
`./mvnw  quarkus:dev`

gives me:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.6.0.Final:dev (default-cli) on project h4kcluster-quarkus: Execution default-cli of goal io.quarkus:quarkus-maven-plugin:1.6.0.Final:dev failed.: NullPointerException 
```

`./mvnw compile quarkus:dev`
doesn't log errors but the application only show an not found page:
```
<html><head><title>Error</title></head><body>Not Found</body></html>
```

Instead
`./mvnw -f launcher quarkus:dev`
works perfectly, without the compile.
It's not a big issue but it's a bit confusing

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -506,13 +506,11 @@ public class DevMojo extends AbstractMojo {
 
         String projectDirectory = null;
         Set<String> sourcePaths = null;
-        Set<String> sourceGenPaths = null;
         String classesPath = null;
         String resourcePath = null;
 
         final MavenProject mavenProject = session.getProjectMap().get(
                 String.format("%s:%s:%s", localProject.getGroupId(), localProject.getArtifactId(), localProject.getVersion()));
-
         if (mavenProject == null) {
             projectDirectory = localProject.getDir().toAbsolutePath().toString();
             Path sourcePath = localProject.getSourcesSourcesDir().toAbsolutePath();
@@ -530,7 +528,6 @@ public class DevMojo extends AbstractMojo {
                     .map(src -> src.toAbsolutePath().toString())
                     .collect(Collectors.toSet());
         }
-
         Path sourceParent = localProject.getSourcesDir().toAbsolutePath();
 
         Path classesDir = localProject.getClassesDir();
@@ -542,7 +539,8 @@ public class DevMojo extends AbstractMojo {
             resourcePath = resourcesSourcesDir.toAbsolutePath().toString();
         }
 
-        Path targetDir = Paths.get(project.getBuild().getOutputDirectory()).getParent();
+        Path targetDir = Paths.get(project.getBuild().getDirectory());
+
         DevModeContext.ModuleInfo moduleInfo = new DevModeContext.ModuleInfo(localProject.getKey(),
                 localProject.getArtifactId(),
                 projectDirectory,
@@ -674,15 +672,14 @@ public class DevMojo extends AbstractMojo {
             }
 
             setKotlinSpecificFlags(devModeContext);
-            final LocalProject localProject;
             if (noDeps) {
-                localProject = LocalProject.load(project.getModel().getPomFile().toPath());
+                final LocalProject localProject = LocalProject.load(project.getModel().getPomFile().toPath());
                 addProject(devModeContext, localProject, true);
                 pomFiles.add(localProject.getRawModel().getPomFile().toPath());
                 devModeContext.getLocalArtifacts()
                         .add(new AppArtifactKey(localProject.getGroupId(), localProject.getArtifactId(), null, "jar"));
             } else {
-                localProject = LocalProject.loadWorkspace(project.getModel().getPomFile().toPath());
+                final LocalProject localProject = LocalProject.loadWorkspace(project.getModel().getPomFile().toPath());
                 for (LocalProject project : filterExtensionDependencies(localProject)) {
                     addProject(devModeContext, project, project == localProject);
                     pomFiles.add(project.getRawModel().getPomFile().toPath());

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java
@@ -12,6 +12,8 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
+import org.apache.maven.model.Build;
 import org.apache.maven.model.Model;
 import org.apache.maven.model.Parent;
 import org.apache.maven.model.Resource;
@@ -25,6 +27,7 @@ public class LocalProject {
     public static final String PROJECT_GROUPID = "${project.groupId}";
 
     private static final String PROJECT_BASEDIR = "${project.basedir}";
+    private static final String PROJECT_BUILD_DIR = "${project.build.directory}";
     private static final String POM_XML = "pom.xml";
 
     private static class WorkspaceLoader {
@@ -265,6 +268,17 @@ public class LocalProject {
         this.version = resolvedVersion;
     }
 
+    public LocalProject getLocalParent() {
+        if (workspace == null) {
+            return null;
+        }
+        final Parent parent = rawModel.getParent();
+        if (parent == null) {
+            return null;
+        }
+        return workspace.getProject(parent.getGroupId(), parent.getArtifactId());
+    }
+
     public String getGroupId() {
         return groupId;
     }
@@ -282,26 +296,23 @@ public class LocalProject {
     }
 
     public Path getOutputDir() {
-        return dir.resolve("target");
+        return resolveRelativeToBaseDir(configuredBuildDir(this, build -> build.getDirectory()), "target");
     }
 
     public Path getCodeGenOutputDir() {
-        return dir.resolve("target").resolve("generated-sources");
+        return getOutputDir().resolve("generated-sources");
     }
 
     public Path getClassesDir() {
-        final String classesDir = rawModel.getBuild() == null ? null : rawModel.getBuild().getOutputDirectory();
-        return resolveRelativeToBaseDir(classesDir, "target/classes");
+        return resolveRelativeToBuildDir(configuredBuildDir(this, build -> build.getOutputDirectory()), "classes");
     }
 
     public Path getTestClassesDir() {
-        final String classesDir = rawModel.getBuild() == null ? null : rawModel.getBuild().getTestOutputDirectory();
-        return resolveRelativeToBaseDir(classesDir, "target/test-classes");
+        return resolveRelativeToBuildDir(configuredBuildDir(this, build -> build.getTestOutputDirectory()), "test-classes");
     }
 
     public Path getSourcesSourcesDir() {
-        final String srcDir = rawModel.getBuild() == null ? null : rawModel.getBuild().getSourceDirectory();
-        return resolveRelativeToBaseDir(srcDir, "src/main/java");
+        return resolveRelativeToBaseDir(configuredBuildDir(this, build -> build.getSourceDirectory()), "src/main/java");
     }
 
     public Path getSourcesDir() {
@@ -337,10 +348,26 @@ public class LocalProject {
     }
 
     private Path resolveRelativeToBaseDir(String path, String defaultPath) {
-        return dir.resolve(path == null ? defaultPath : stripProjectBasedirPrefix(path));
+        return dir.resolve(path == null ? defaultPath : stripProjectBasedirPrefix(path, PROJECT_BASEDIR));
     }
 
-    private static String stripProjectBasedirPrefix(String path) {
-        return path.startsWith(PROJECT_BASEDIR) ? path.substring(PROJECT_BASEDIR.length() + 1) : path;
+    private Path resolveRelativeToBuildDir(String path, String defaultPath) {
+        return getOutputDir().resolve(path == null ? defaultPath : stripProjectBasedirPrefix(path, PROJECT_BUILD_DIR));
+    }
+
+    private static String stripProjectBasedirPrefix(String path, String expr) {
+        return path.startsWith(expr) ? path.substring(expr.length() + 1) : path;
+    }
+
+    private static String configuredBuildDir(LocalProject project, Function<Build, String> f) {
+        String dir = project.rawModel.getBuild() == null ? null : f.apply(project.rawModel.getBuild());
+        while (dir == null) {
+            project = project.getLocalParent();
+            if (project == null) {
+                break;
+            }
+            dir = project.rawModel.getBuild() == null ? null : f.apply(project.rawModel.getBuild());
+        }
+        return dir;
     }
 }

--- a/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
+++ b/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
@@ -7,6 +7,7 @@ import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
 
 import io.quarkus.bootstrap.model.AppArtifactKey;
 import io.quarkus.bootstrap.resolver.maven.workspace.LocalProject;
@@ -125,6 +126,7 @@ public class LocalWorkspaceDiscoveryTest {
 
         assertEquals("acme", project.getArtifactId());
         assertWorkspaceWithParentInChildDir(project);
+        assertParents(project, "acme-parent", "acme-dependencies");
     }
 
     @Test
@@ -138,6 +140,8 @@ public class LocalWorkspaceDiscoveryTest {
 
         assertEquals("acme-application", project.getArtifactId());
         assertWorkspaceWithParentInChildDir(project);
+
+        assertParents(project, "acme-parent", "acme-dependencies");
     }
 
     private void assertWorkspaceWithParentInChildDir(final LocalProject project) {
@@ -159,13 +163,20 @@ public class LocalWorkspaceDiscoveryTest {
         final Path projectDir = Paths.get(projectUrl.toURI());
         assertTrue(Files.exists(projectDir));
         final LocalProject project = LocalProject.loadWorkspace(projectDir);
+        assertParents(project, "root");
 
         assertEquals("root-module1", project.getArtifactId());
         final LocalWorkspace workspace = project.getWorkspace();
-        assertNotNull(workspace.getProject("org.acme", "root"));
-        assertNotNull(workspace.getProject("org.acme", "root-module1"));
+        LocalProject rootProject = workspace.getProject("org.acme", "root");
+        assertNotNull(rootProject);
+        assertNull(rootProject.getLocalParent());
+        LocalProject module1 = workspace.getProject("org.acme", "root-module1");
+        assertNotNull(module1);
+        assertParents(module1, "root");
         assertNull(workspace.getProject("org.acme", "root-module2"));
-        assertNotNull(workspace.getProject("org.acme", "root-submodule"));
+        LocalProject submodule = workspace.getProject("org.acme", "root-submodule");
+        assertNotNull(submodule);
+        assertParents(submodule, "root-module1", "root");
         assertEquals(3, workspace.getProjects().size());
     }
 
@@ -177,12 +188,19 @@ public class LocalWorkspaceDiscoveryTest {
         final Path projectDir = Paths.get(projectUrl.toURI());
         assertTrue(Files.exists(projectDir));
         final LocalProject project = LocalProject.loadWorkspace(projectDir);
+        assertParents(project, "root");
 
         assertEquals("root-module1", project.getArtifactId());
         final LocalWorkspace workspace = project.getWorkspace();
-        assertNotNull(workspace.getProject("org.acme", "root"));
-        assertNotNull(workspace.getProject("org.acme", "root-module1"));
-        assertNotNull(workspace.getProject("org.acme", "root-module2"));
+        LocalProject root = workspace.getProject("org.acme", "root");
+        assertNotNull(root);
+        assertNull(root.getLocalParent());
+        LocalProject module1 = workspace.getProject("org.acme", "root-module1");
+        assertNotNull(module1);
+        assertParents(module1, "root");
+        LocalProject module2 = workspace.getProject("org.acme", "root-module2");
+        assertNotNull(module2);
+        assertParents(module2, "root");
         assertNull(workspace.getProject("org.acme", "root-submodule"));
         assertEquals(3, workspace.getProjects().size());
     }
@@ -199,6 +217,8 @@ public class LocalWorkspaceDiscoveryTest {
         final Map<AppArtifactKey, LocalProject> projects = project.getWorkspace().getProjects();
         assertEquals(6, projects.size());
         assertTrue(projects.containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "independent")));
+
+        assertNull(project.getLocalParent());
     }
 
     @Test
@@ -210,6 +230,7 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(MvnProjectBuilder.DEFAULT_GROUP_ID, project.getGroupId());
         assertEquals("root-no-parent-module", project.getArtifactId());
         assertEquals(MvnProjectBuilder.DEFAULT_VERSION, project.getVersion());
+        assertNull(project.getLocalParent());
     }
 
     @Test
@@ -224,6 +245,8 @@ public class LocalWorkspaceDiscoveryTest {
         final Map<AppArtifactKey, LocalProject> projects = project.getWorkspace().getProjects();
         assertEquals(5, projects.size());
         assertTrue(projects.containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "root-no-parent-module")));
+
+        assertParents(project);
     }
 
     @Test
@@ -235,6 +258,8 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(MvnProjectBuilder.DEFAULT_GROUP_ID, project.getGroupId());
         assertEquals("root-module-with-parent", project.getArtifactId());
         assertEquals(MvnProjectBuilder.DEFAULT_VERSION, project.getVersion());
+
+        assertParents(project);
     }
 
     @Test
@@ -248,6 +273,7 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(MvnProjectBuilder.DEFAULT_VERSION, project.getVersion());
 
         assertCompleteWorkspace(project);
+        assertParents(project, "root");
     }
 
     @Test
@@ -261,6 +287,7 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(MvnProjectBuilder.DEFAULT_VERSION, project.getVersion());
 
         assertCompleteWorkspace(project);
+        assertParents(project, "root");
     }
 
     @Test
@@ -281,6 +308,8 @@ public class LocalWorkspaceDiscoveryTest {
         assertTrue(projects.containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "another-child")));
         assertTrue(projects
                 .containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "empty-parent-relative-path-module")));
+
+        assertParents(project, "root");
     }
 
     /**
@@ -297,6 +326,8 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(MvnProjectBuilder.DEFAULT_VERSION, project.getVersion());
 
         assertCompleteWorkspace(project);
+
+        assertParents(project);
     }
 
     @Test
@@ -353,6 +384,28 @@ public class LocalWorkspaceDiscoveryTest {
         testMavenCiFriendlyVersion("${revision}${sha1}${changelist}", "workspace-multiple", expectedResolvedVersion, false);
     }
 
+    @Test
+    public void testBuildDirs() throws Exception {
+        final URL projectUrl = Thread.currentThread().getContextClassLoader()
+                .getResource("build-directories/multimodule/runner");
+        assertNotNull(projectUrl);
+        final Path runnerDir = Paths.get(projectUrl.toURI());
+        assertTrue(Files.exists(runnerDir));
+        final LocalProject project = LocalProject.loadWorkspace(runnerDir);
+        assertNotNull(project);
+        assertEquals(runnerDir.resolve("custom-target"), project.getOutputDir());
+        assertEquals(runnerDir.resolve("src/main/other"), project.getSourcesSourcesDir());
+        assertEquals(runnerDir.resolve("custom-target").resolve("other-classes"), project.getClassesDir());
+        assertEquals(runnerDir.resolve("custom-target").resolve("test-classes"), project.getTestClassesDir());
+
+        final LocalProject parent = project.getLocalParent();
+        final Path parentDir = parent.getDir();
+        assertEquals(parentDir.resolve("custom-target"), parent.getOutputDir());
+        assertEquals(parentDir.resolve("src/main/other"), parent.getSourcesSourcesDir());
+        assertEquals(parentDir.resolve("custom-target").resolve("custom-classes"), parent.getClassesDir());
+        assertEquals(parentDir.resolve("custom-target").resolve("test-classes"), parent.getTestClassesDir());
+    }
+
     private void testMavenCiFriendlyVersion(String placeholder, String testResourceDirName, String expectedResolvedVersion,
             boolean resolvesFromWorkspace) throws Exception {
         final URL module1Url = Thread.currentThread().getContextClassLoader()
@@ -391,5 +444,20 @@ public class LocalWorkspaceDiscoveryTest {
         assertTrue(projects
                 .containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "empty-parent-relative-path-module")));
         assertTrue(projects.containsKey(new AppArtifactKey(MvnProjectBuilder.DEFAULT_GROUP_ID, "root")));
+    }
+
+    private static void assertParents(LocalProject project, String... parentArtifactId) {
+        LocalProject parent = project.getLocalParent();
+        int i = 0;
+        while (parent != null) {
+            if (i == parentArtifactId.length) {
+                fail("Unexpected parent " + parent.getKey());
+            }
+            assertEquals(parentArtifactId[i++], parent.getArtifactId());
+            parent = parent.getLocalParent();
+        }
+        if (i != parentArtifactId.length) {
+            fail("Missing parent " + parentArtifactId[i]);
+        }
     }
 }

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/build-directories/multimodule/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/build-directories/multimodule/pom.xml
@@ -1,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.acme</groupId>
+    <artifactId>multimodule-parent</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+
+    <properties>
+    </properties>
+
+    <modules>
+        <module>runner</module>
+    </modules>
+
+    <build>
+        <directory>custom-target</directory>
+        <outputDirectory>${project.build.directory}/custom-classes</outputDirectory>
+        <sourceDirectory>src/main/other</sourceDirectory>
+        <testSourceDirectory>src/test/other</testSourceDirectory>
+    </build>
+
+    <dependencyManagement>
+    </dependencyManagement>
+
+</project>

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/build-directories/multimodule/runner/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/build-directories/multimodule/runner/pom.xml
@@ -1,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.acme</groupId>
+        <artifactId>multimodule-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../</relativePath>
+    </parent>
+    <artifactId>multimodule-main</artifactId>
+
+    <dependencies>
+    </dependencies>
+
+    <build>
+        <outputDirectory>${project.build.directory}/other-classes</outputDirectory>
+    </build>
+</project>

